src/net_processing.cpp

```diff
@@ -3731,24 +3838,44 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
         //
         // Message: getdata (non-blocks)
         //
-        while (!pto->mapAskFor.empty() && (*pto->mapAskFor.begin()).first <= nNow)
+        std::multimap<int64_t, uint256> &tx_process_time = state.m_tx_download.m_tx_process_time;
+        while (!tx_process_time.empty() && (*tx_process_time.begin()).first <= nNow && state.m_tx_download.m_in_flight.size() < MAX_PEER_TX_IN_FLIGHT)
         {
-            const CInv& inv = (*pto->mapAskFor.begin()).second;
+            const uint256& txid = (*tx_process_time.begin()).second;
+            CInv inv(MSG_TX | GetFetchFlags(pto), txid);
             if (!AlreadyHave(inv))
             {
-                LogPrint(BCLog::NET, "Requesting %s peer=%d\n", inv.ToString(), pto->GetId());
-                vGetData.push_back(inv);
-                if (vGetData.size() >= 1000)
-                {
-                    connman->PushMessage(pto, msgMaker.Make(NetMsgType::GETDATA, vGetData));
-                    vGetData.clear();
+                // If this transaction was last request more than 1 minute ago,
+                // then request.
+                int64_t last_request_time = GetTxRequestTime(inv.hash);
+                if (last_request_time <= nNow - MAX_GETDATA_TX_DELAY) {
+                    LogPrint(BCLog::NET, "Requesting %s peer=%d\n", inv.ToString(), pto->GetId());
+                    vGetData.push_back(inv);
+                    if (vGetData.size() >= 1000)
+                    {
+                        connman->PushMessage(pto, msgMaker.Make(NetMsgType::GETDATA, vGetData));
+                        vGetData.clear();
+                    }
+                    UpdateTxRequestTime(inv.hash, nNow);
+                    state.m_tx_download.m_in_flight.insert(inv.hash);
+                } else {
+                    // This transaction is in flight from someone else; queue
+                    // up processing to happen after the download times out
+                    // (with a slight delay for inboudnd peers, to prefer
+                    // requests to outbound peers).
+                    int64_t new_process_time = last_request_time + MAX_GETDATA_TX_DELAY;
```

Comment:
Could this just call `RequestTx`? To avoid spreading the INBOUND_PEER_TX_DELAY logic over two places in the code.